plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/TypeEmbedding.kt

```diff
@@ -251,5 +261,49 @@ data class ClassTypeEmbedding(val className: ScopedKotlinName) : TypeEmbedding {
     override fun accessInvariants(v: Exp): List<Exp> =
         flatMapUniqueFields { _, field -> field.accessInvariantsForParameter(v) }
 
+    // Note: this function will replace accessInvariants when nested unfold will be implemented
+    override fun predicateAccessInvariants(v: Exp): List<Exp> = listOf(Exp.PredicateAccess(name, listOf(v)))
+
     override fun provenInvariants(v: Exp) = listOf(subtypeInvariant(v))
+
+    private val thisEmbedding = VariableEmbedding(ThisReceiverName, this)
+    val classPredicate: Predicate
+        get() {
+            val accessReadableFields = fields.values
+                .filter { it.accessPolicy == AccessPolicy.ALWAYS_READABLE }
+                .map { FieldAccess(thisEmbedding, it).getAccessPredicate(PermExp.WildcardPerm()) }
+
+            val accessSpecialFields = fields.values
+                .filter { it.accessPolicy == AccessPolicy.ALWAYS_WRITEABLE }
+                .map { FieldAccess(thisEmbedding, it).getAccessPredicate(PermExp.FullPerm()) }
+
+            val classFieldPredicates = fields.values
+                .filter { it.accessPolicy == AccessPolicy.ALWAYS_READABLE }
```

Comment:
Is the reason for the `ALWAYS_READABLE` here because writeable fields can't be accessed at this point?

Reply:
It is still not clear to me whether `ALWAYS_WRITEABLE` are just special fields or might be something else. For the moment I think that the only `ALWAYS_WRITEABLE` field is `size`, so it makes no difference include them here or not (because `size` is of type `Int`). Let me know what's the best thing to do for you.

Reply:
I am not sure, to be honest.  Could you leave a comment that says that this is something we should evaluate?  I think allowing this for `ALWAYS_WRITEABLE` would be asking for trouble, but at the same time, if something is always writeable, we should also always have permissions for it...